### ISSUE TITLE
Incorrect output shape for max-pooling

### ISSUE THREAD

**Issue by f0k** [pytorch/pytorch#13386]
## 🐛 Bug

For some combinations of input size, pooling size and padding, the output size of `max_pool1d` is wrong. Furthermore, it is different between CPU and CUDA.

## To Reproduce

On GPU:
```python
import torch
x = torch.rand(19200000 - 2, dtype=torch.float32, device='cuda:0')
width = 1921
y = torch.nn.functional.max_pool1d(x[None, None], width, stride=1, padding=width//2)[0, 0]
print(x.shape)
print(y.shape)
```
I get:
```
torch.Size([19199998])
torch.Size([19199996])
```

On CPU:
```python
import torch
x = torch.rand(19200000 - 2, dtype=torch.float32, device='cpu')
width = 1921
y = torch.nn.functional.max_pool1d(x[None, None], width, stride=1, padding=width//2)[0, 0]
print(x.shape)
print(y.shape)
```
I get:
```
torch.Size([19199998])
torch.Size([19199997])
```

For an input length of `19200000`, I get an output length of `19200000` on GPU and `19200001` on CPU.

## Expected behavior

In all cases, the output shape should match the input shape -- it's an uneven pooling window with stride 1 and matching symmetric padding. Also the output shape should be the same regardless of the device. Note that it works fine on both CUDA and CPU for an input length of `1920000 - 2` (one order of magnitude smaller).

## Environment

I can reproduce this both with a precompiled PyTorch 0.4.1 and a self-compiled PyTorch 1.0.

Pre-compiled:
```
Collecting environment information...
PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.4 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 2.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: TITAN X (Pascal)
GPU 1: TITAN X (Pascal)

Nvidia driver version: 390.30
cuDNN version: Probably one of the following:
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn.so
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn.so.5
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn.so.5.1.10
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn_static.a

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect
```

Self-compiled:
```
PyTorch version: 1.0.0a0+710191e
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 2.7
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: GeForce GTX 1060
Nvidia driver version: 410.57
cuDNN version: Probably one of the following:
/usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudnn.so.7.3.1
/usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudnn_static.a
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn.so.7.1.4
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn_static.a

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect
```

## Additional context

For what it's worth, it works fine using Lasagne/Theano both on CPU and CUDA.

**Comment by f0k** [comment]
## Follow-up

It's the same with e475d3e. And I have a horrifying suspicion what you're doing:
```
In [1]: inp = 19199998
In [2]: pad = 1921//2
In [3]: dil = 1
In [4]: krn = 1921
In [5]: std = 1
In [6]: one = 1
In [7]: two = 2
In [8]: (inp + two*pad - dil*(krn-one) - one) // std + one
Out[8]: 19199998
In [9]: inp, pad, dil, krn, std, one, two = map(np.float32, (inp, pad, dil, krn, std, one, two))
In [10]: (inp + two*pad - dil*(krn-one) - one) // std + one
Out[10]: 19199996.0
```
So my theory is you're using floating point for computing the output shape. This wouldn't explain the CPU/GPU difference, but it would explain why the size can be off in both directions, and only for large inputs.

This should be a blocker for 1.0.

**Comment by colesbury** [comment]
Yup... looks like we're using floating point calculation:

https://github.com/pytorch/pytorch/blob/26a8bb62ee359d3fa444dc55133f11a4a42f0479/aten/src/THNN/generic/SpatialDilatedMaxPooling.c#L45-L51

**Comment by zou3519** [comment]
cc @nairbv

**Comment by f0k** [comment]
> Yup... looks like we're using floating point calculation:

Thanks for pointing out the source, I must admit that I was a bit lost.

For the GPU, the conversion to float is placed slightly differently, that's why the results do not always match up:
https://github.com/pytorch/pytorch/blob/26a8bb62ee359d3fa444dc55133f11a4a42f0479/aten/src/THCUNN/generic/SpatialDilatedMaxPooling.cu#L48-L53

> This problem also occurs for `avg_pool1d`:

All pooling operations will be affected, because somebody thought that's the way to distinguish the `ceil`/`floor` cases.

Convolution does not have a `ceil` mode, so it's done in int64 as it should:
https://github.com/pytorch/pytorch/blob/26a8bb62ee359d3fa444dc55133f11a4a42f0479/aten/src/THNN/generic/SpatialDilatedConvolution.c#L243-L244

So the actual fix will be easy, I'd be happy to do it, something like:
```C
int64_t outp = (inp + 2*pad - dilation * (kernel - 1) - 1);
outp = outp / stride + 1 + (ceilmode && (outp % stride) ? 1 : 0);
```
But I also need to learn where to add a corresponding test.

**Comment by nairbv** [comment]
@f0k you're welcome to take a shot at it, I think a test could probably go in test/test_nn.py.
